src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobTask.cs

```diff
@@ -766,17 +773,32 @@ private async Task<(int totalCount, List<string> resourcesTypes)> CalculateTotal
             return (totalCount, resourcesTypes);
         }
 
-        private async Task UpdateParametersAndCompleteJob()
+        private async Task UpdateParametersAndCompleteJob(CancellationToken cancellationToken)
         {
             // here we check if all the resource types which are base types of the search parameter
             // were reindexed by this job.  If so, then we should mark the search parameters
             // as fully reindexed
             var fullyIndexedParamUris = new List<string>();
             var reindexedResourcesSet = new HashSet<string>(_reindexJobRecord.Resources);
+            var searchParamStatusCollection = await _searchParameterStatusManager.GetAllSearchParameterStatus(cancellationToken);
             foreach (var searchParam in _reindexJobRecord.SearchParams)
             {
-                var searchParamInfo = _supportedSearchParameterDefinitionManager.GetSearchParameter(searchParam);
-                if (reindexedResourcesSet.IsSupersetOf(searchParamInfo.BaseResourceTypes))
+                // go to base search param definition manager
+                var searchParamInfo = _searchParameterDefinitionManager.GetSearchParameter(searchParam);
+                var spStatus = searchParamStatusCollection.FirstOrDefault(sp => sp.Uri.Equals(searchParamInfo.Url)).Status;
+
+                // Adding the below to explicitly update from their pending states to their final states.
+                if (spStatus == Search.Registry.SearchParameterStatus.PendingDisable)
+                {
+                    _logger.LogInformation("Reindex job updating the status of the fully indexed search, id: {Id}, parameter: '{ParamUri}' to Disabled.", _reindexJobRecord.Id, searchParam);
+                    await _searchParameterStatusManager.UpdateSearchParameterStatusAsync(new List<string>() { searchParamInfo.Url.ToString() }, SearchParameterStatus.Disabled, cancellationToken);
+                }
+                else if (spStatus == Search.Registry.SearchParameterStatus.PendingDelete)
+                {
+                    _logger.LogInformation("Reindex job updating the status of the fully indexed search, id: {Id}, parameter: '{ParamUri}' to Deleted.", _reindexJobRecord.Id, searchParam);
+                    await _searchParameterStatusManager.UpdateSearchParameterStatusAsync(new List<string>() { searchParamInfo.Url.ToString() }, SearchParameterStatus.Deleted, cancellationToken);
+                }
+                else if (reindexedResourcesSet.IsSupersetOf(searchParamInfo.BaseResourceTypes) && (spStatus == SearchParameterStatus.Supported || spStatus == SearchParameterStatus.Enabled))
```

Comment:
Do we need a similar supserset check on the pendingDisable or PendingDelete search parameters, in case the reindex job was run for only a subset of the resource types for the pending SearchParameters?

Reply:
This change is just to make sure that the pending disable and pending delete don't get set to enabled, which was happening previously. This just makes sure they don't get included in the list for search parameter updates to a status of enabled. Maybe I'm not understanding what you are asking.

Reply:
On line 801, we only update supported -> Enabled if the resource types that were reindexed, were the complete list of resource types the search parameter applied to.

But on lines 791 and 796, we just check the status of the search parameter being in a pending state.  It's possible that we have a SearchParameter X  which applies to resource types 1,2 and 3.  Then we set X to pendingDisable.  Then we run a reindex job for only resource type 2.  resource types 1 and 3 still need to be cleaned up, but it hasn't happened yet.  We don't want to mark SearchParameter X to Disabled, because it's not really finished yet.  We should leave it as pending.

Reply:
I see, that makes sense. Let me look into that as we should have the same type of check.

Reply:
Ok, I moved it to the beginning, and only update any of the statuses if that checks out first. Thanks for the catch.